content/blog/new-per-subject-discard-policy.md

```diff
@@ -0,0 +1,223 @@
+
+So by default you can say that the 'per subject discard policy' is `old`, regardless of the stream's discard policy being `new` or `old`. 
+
+This is because it is the behavior you want for one of the most basic and common use cases of streaming: the ability to act as a 'last message cache' so that subscribing applications immediately get an 'initial value' (the current latest message published) for the subject(s) they are subscribing to, rather than having to wait an unknown period of time until the next publication on that (or these) subjects to get any data.
+
+### Using Publish as an 'Insert'
+
+However, and this is the new feature being introduced in 2.9, JetStream also allows you to create streams with the `new-per-subject` attribute which sets the _per subject_ discard policy to `new` when the stream's discard policy is also set to `new`.
+
+So for example if you have set the steam's max messages per subject limit to 1 and there is already a message stored in the stream with that subject then the JetStream publish call fails immediately, just like an `insert` fails if there's already a row for that key in a table with a unique id constraint in SQL.
+
+This in turns gives us infinite exactly-once publication quality of service, simply by including the unique id somewhere in the subject name.
+
+### Using Publish as an 'Update'
+
+While it is outside the subject of this blog, I still want to mention that JetStream also has a 'compare and set' functionality (using the `nats.ExpectLastSequence()` publish option) which can be used to implement an equivalent to `update` in SQL.
+
+## How does it work?
+The ability to strictly and consistently enforce this limit (a key attribute for our use case) is possible because of a fundamental difference between JetStream and distributed log streaming systems: JetStream is 'subject-based addressing aware', it allows you to capture messages on multiple subjects into a single stream, rather than having a hard equivalency of exactly one topic per stream. Of course other streaming systems also allow the publishing application to provide a 'key' field along with the message being published to a topic, but the key is really meant to be only used for distribution (partitioning) purposes. But that is not the same thing as having the key(s) in the subject and then being able to use subject-based server-side (indexed) filtering consumers of the messages in a stream.
+
+### Doesn't Kafka do this also?
+And indeed, if you know your Kafka, this may look to you just like the 'keep at least one message per key' log compaction feature, but there is a big difference: JetStream is 'exactly *n*' maximum number of messages per subject while log compaction only ensures that you will have 'at least 1' message per key.
```

Comment:
```suggestion
And indeed, if you have experience with Kafka, this may look to you just like the "keep at least one message per key" log compaction feature, but there is a big difference: JetStream is exactly *n* maximum number of messages per subject while log compaction only ensures that you will have *at least one* message per key.
```